src/backend/distributed/metadata/dependency.c

```diff
@@ -787,6 +788,47 @@ EnsureRelationDependenciesCanBeDistributed(ObjectAddress *relationAddress)
 		ereport(ERROR, (errmsg("Relation \"%s\" has dependency on unsupported "
 							   "object \"%s\"", tableName, objectType)));
 	}
+
+	/*
+	 * Having circular dependency between distributed objects prevents Citus from
+	 * adding a new node. So, error out while adding a new obejct if addition of
+	 * that object causes a circular dependency
+	 */
+	ErrorIfCircularDependencyExists(relationAddress);
```

Comment:
is there a more generic spot where we can put this to also cover cycles that do not involve relations?

Reply:
I've moved that check to EnsureDependenciesExistOnAllNodes.